master/internal/rm/kubernetesrm/informer_intg_test.go

```diff
@@ -0,0 +1,228 @@
+	mockOptsWatch := metaV1.ListOptions{
+		LabelSelector:       determinedLabel,
+		ResourceVersion:     "1",
+		AllowWatchBookmarks: true,
+	}
+
+	mockPod := &mocks.PodInterface{}
+
+	mockPod.On("List", ctx, mockOptsList).Return(
+		&k8sV1.PodList{
+			ListMeta: metaV1.ListMeta{
+				ResourceVersion: "1",
+			}},
+		nil)
+	mockPod.On("Watch", ctx, mockOptsWatch).Return(&mockWatcher{c: c}, nil)
+
+	return mockPod
+}
+
+func initializeMockNodeInterface(c chan watch.Event) *mocks.NodeInterface {
+	ctx := context.TODO()
+	mockOptsList := metaV1.ListOptions{LabelSelector: determinedLabel}
+	mockOptsWatch := metaV1.ListOptions{
+		LabelSelector:       determinedLabel,
+		ResourceVersion:     "1",
+		AllowWatchBookmarks: true,
+	}
+
+	mockNode := &mocks.NodeInterface{}
+
+	mockNode.On("List", ctx, mockOptsList).Return(
+		&k8sV1.NodeList{
+			ListMeta: metaV1.ListMeta{
+				ResourceVersion: "1",
+			}},
+		nil)
+	mockNode.On("Watch", ctx, mockOptsWatch).Return(&mockWatcher{c: c}, nil)
+
+	return mockNode
+}
+
+func TestStartInformer(t *testing.T) {
+	cases := []struct {
+		name            string
+		testNamespace   string
+		expected        error
+		podNames        []string
+		orderedPodNames []string
+	}{
+		{"informer success", namespace,
+			nil, []string{"abc"}, []string{"abc"}},
+		{"informer success & event ordering success", namespace,
+			nil, []string{"A", "B", "C", "D", "E"},
+			[]string{"A", "B", "C", "D", "E"}},
+		{"informer success & event ordering failure", namespace,
+			nil, []string{"A", "B", "C", "D", "E"},
+			[]string{"A", "A", "C", "E", "D"}},
+		{"podInterface failure", "NOT",
+			errors.New("newInformer: passed podInterface is nil"),
+			[]string{}, []string{}},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			var wg sync.WaitGroup
+			eventChan := make(chan watch.Event)
+			ordering := make([]string, 0)
+			mockPodInterface := initializeMockPodInterface(eventChan)
+
+			mockPods := &pods{
+				namespace:     namespace,
+				podInterfaces: map[string]typedV1.PodInterface{namespace: mockPodInterface},
+			}
+			mockPodHandler := func(pod *k8sV1.Pod) {
+				t.Logf("received pod %v", pod)
+				ordering = append(ordering, pod.Name)
+				wg.Done()
+			}
+
+			wg.Add(len(tt.podNames))
+
+			// Test newInformer.
+			i, err := newInformer(context.TODO(), tt.testNamespace,
+				mockPods.podInterfaces[tt.testNamespace])
+			if err != nil {
+				assert.Nil(t, i)
+				assert.Error(t, tt.expected, err)
+				return
+			}
+			assert.NotNil(t, i)
+			assert.Equal(t, tt.expected, err)
+
+			// Test startInformer & assert correct ordering of pod-modified events.
+			go i.startInformer(mockPodHandler)
+			for _, name := range tt.podNames {
+				pod := &k8sV1.Pod{
+					ObjectMeta: metaV1.ObjectMeta{
+						ResourceVersion: "1",
+						Name:            name,
+					}}
+				eventChan <- watch.Event{
+					Type:   watch.Modified,
+					Object: pod,
+				}
+			}
```

Comment:
I had to add the `close(eventChan)` and `wg.Wait()` to the go func() to avoid panicking, but I think it still achieves the same result.

Reply:
actually, that implemention didn't pass the test, so I reverted back to:
```
go i.startInformer()
...
wg.Wait()
asser.Equal(t, tt.podNames, ordering)
```
I think the `defer wg.Done()` never executes in the goroutine because once the Informer starts, it never returns anything (or stops). Happy to take another look at it, but the current proposal panics :(

Reply:
Sorry for sending you on a wild goose chase! I tried this myself (which is what I should have done in the first place before proposing it) to figure out what the panics were about and it looks like the `RetryWatcher` closes its own channel when it shuts down, so we shouldn't do that or we'll get a "close of closed channel" panic.

Plus, if the expectation (as implied [here](https://github.com/determined-ai/determined/blob/2973a3048e442fd3cbfe07cf68aeb82d8dfdbfcb/master/internal/rm/kubernetesrm/nodes.go#L70)) is that the informer will run forever once started, it probably doesn't make as much sense as I thought to shut it down as part of its testing.

I did come up with [a way to make this work](https://github.com/determined-ai/determined/commit/b8aac2b578b4fc884cff33da703956be3a567c30), but it involves some extra `Context` logic (i.e. making `startNodeInformer` stoppable via canceling a the `Context` passed into `newNodeInformer`) so I'm not sure it's worth it since we don't really expect `nodeInformer` to ever stop. I'll leave that call up to you. Either way, this looks good as is to me, so I'm going to mark this as resolved.